automerge/src/transaction/inner.rs

```diff
@@ -184,13 +198,40 @@ impl TransactionInner {
         index: usize,
         action: OpType,
     ) -> Result<Option<OpId>, AutomergeError> {
-        let id = self.next_id();
-
-        let query = doc.ops.search(&obj, query::InsertNth::new(index));
+        let key = if let Some(buffer) = self.insert_buffer.as_mut() {
```

Comment:
What if, instead of an `Option<InputBuffer>`, the `InputBuffer` was an enum with an empty state, e.g:

```rust
enum InputBuffer{
    Empty,
    Buffer {
        /// object that these inserts are operating on
        obj: ObjId,
        /// index that the first op in this buffer wanted to insert at
        target_index: usize,
        /// number of actions this buffer has seen
        num_actions: usize,
        /// index in the optree that the first op in this buffer was inserted in
        tree_index: usize,
    }
}
```
If we do this and move the logic here into an `InputBuffer::next_op` then I think you would be able to remove the `unwrap` below and this whole chunk of code would become something like:

```rust
doc.ops
            .insert(buffer.next_op(&mut this.ops, obj, index, action));
```